File: sdk/typescript/src/models/layered-config.ts
import { normalizeSnapshotCommands } from "../snapshot-commands";

export interface DriveSpec {
  drive_id: string;
  label?: string;
  size_gb?: number;
  mount_path?: string;
}

export interface LayerDef {
  name: string;
  init_commands: Array<Record<string, unknown>>;
  refresh_commands?: Array<Record<string, unknown>>;
  drives?: Array<DriveSpec>;
  refresh_interval?: string;
}

export interface LayeredConfigConfig {
  auto_pause?: boolean;
  ttl?: number;
  tier?: string;
  auto_rollout?: boolean;
  session_max_age_seconds?: number;
  rootfs_size_gb?: number;
  runner_user?: string;
  workspace_size_gb?: number;
  network_policy_preset?: string;
  network_policy?: unknown;
  auth?: unknown;
}

export interface StoredLayeredConfig {
  config_id: string;
  display_name?: string;
  leaf_layer_hash?: string;
  leaf_workload_key?: string;
  tier?: string;
  start_command?: Record<string, unknown>;
  runner_ttl_seconds?: number;
  session_max_age_seconds?: number;
  auto_pause?: boolean;
  auto_rollout?: boolean;
  max_concurrent_runners?: number;
  build_schedule?: string;
  network_policy_preset?: string;
  network_policy?: unknown;
  created_at?: string;
  updated_at?: string;
}

export interface LayerStatus {
  name: string;
  layer_hash?: string;
  status?: string;
  current_version?: string;
  depth?: number;
  build_status?: string;
  build_version?: string;
}

export interface LayeredConfigDetail {
  config: StoredLayeredConfig;
  layers?: Array<LayerStatus>;
  definition?: unknown;
}

export interface CreateConfigResponse {
  config_id: string;
  leaf_workload_key?: string;
  layers?: Array<Record<string, unknown>>;
}

export interface BuildResponse {
  config_id: string;
  status?: string;
  force?: string;
  clean?: string;
}

export interface RefreshResponse {
  config_id: string;
  layer_name?: string;
  status?: string;
}

export interface LayeredConfigListResponse {
  configs: Array<StoredLayeredConfig>;
  count?: number;
}

export function normalizeLayerDef(layer: LayerDef): LayerDef {
  return {
    ...layer,
    init_commands: normalizeSnapshotCommands(layer.init_commands) ?? [],
    refresh_commands:
      normalizeSnapshotCommands(layer.refresh_commands as
        | Array<string | Record<string, unknown>>
        | undefined),
  };
}

File: sdk/python/src/capsule_sdk/models/layered_config.py
from __future__ import annotations

from typing import Any, cast

from pydantic import Field, field_validator

from capsule_sdk._snapshot_commands import normalize_snapshot_commands
from capsule_sdk.models.common import CapsuleModel


def _empty_configs() -> list[StoredLayeredConfig]:
    return []


class DriveSpec(CapsuleModel):
    """Drive specification for a layer."""

    drive_id: str
    label: str | None = None
    size_gb: int | None = None
    mount_path: str | None = None


class LayerDef(CapsuleModel):
    """Definition of a single layer in a layered config."""

    name: str
    init_commands: list[dict[str, Any]]
    refresh_commands: list[dict[str, Any]] | None = None
    drives: list[DriveSpec] | None = None
    refresh_interval: str | None = None

    @field_validator("init_commands", "refresh_commands", mode="before")
    @classmethod
    def _normalize_commands(cls, value: Any) -> Any:
        if value is None:
            return None
        if not isinstance(value, list):
            return value
        commands = cast(list[str | dict[str, Any]], value)
        return normalize_snapshot_commands(commands)


class LayeredConfigConfig(CapsuleModel):
    """Runtime configuration nested inside a LayeredConfig."""

    auto_pause: bool | None = None
    ttl: int | None = None
    tier: str | None = None
    auto_rollout: bool | None = None
    session_max_age_seconds: int | None = None
    rootfs_size_gb: int | None = None
    runner_user: str | None = None
    workspace_size_gb: int | None = None
    network_policy_preset: str | None = None
    network_policy: Any | None = None
    auth: Any | None = None


class StoredLayeredConfig(CapsuleModel):
    """Server-side stored representation of a layered config."""

    config_id: str
    display_name: str | None = None
    leaf_layer_hash: str | None = None
    leaf_workload_key: str | None = None
    tier: str | None = None
    start_command: dict[str, Any] | None = None
    runner_ttl_seconds: int | None = None
    session_max_age_seconds: int | None = None
    auto_pause: bool | None = None
    auto_rollout: bool | None = None
    max_concurrent_runners: int | None = None
    build_schedule: str | None = None
    network_policy_preset: str | None = None
    network_policy: Any | None = None
    created_at: str | None = None
    updated_at: str | None = None


class LayerStatus(CapsuleModel):
    """Status of a single layer in a layered config."""

    name: str
    layer_hash: str | None = None
    status: str | None = None
    current_version: str | None = None
    depth: int | None = None
    build_status: str | None = None
    build_version: str | None = None


class LayeredConfigDetail(CapsuleModel):
    """Detailed view of a layered config (GET /layered-configs/{id})."""

    config: StoredLayeredConfig
    layers: list[LayerStatus] | None = None
    definition: Any | None = None


class CreateConfigResponse(CapsuleModel):
    """Response from creating a layered config."""

    config_id: str
    leaf_workload_key: str | None = None
    layers: list[dict[str, Any]] | None = None


class BuildResponse(CapsuleModel):
    """Response from triggering a build."""

    config_id: str
    status: str | None = None
    force: bool | None = None
    clean: bool | None = None
    enqueued: int | None = None


class RefreshResponse(CapsuleModel):
    """Response from triggering a layer refresh."""

    config_id: str
    layer_name: str | None = None
    status: str | None = None


class LayeredConfigListResponse(CapsuleModel):
    """Response from listing layered configs."""

    configs: list[StoredLayeredConfig] = Field(default_factory=_empty_configs)
    count: int | None = None
